### ISSUE TITLE
aarch64 AMIs

### ISSUE THREAD

**Issue by tomalok** [mcrute/alpine-ec2-ami#28]
With the advent of a1 (arm64) instance types, we should also build AMIs for them, too.

**Comment by tomalok** [comment]
I managed to get an aarch64 version of `linux-virt` in for alpine 3.9 -- but there's more work yet to do.  Currently at the point where we'll need to use `grub` instead of `syslinux` as a boot loader; there may need to be other as-of-yet unknown arch-specific tweaks needed, too.

**Comment by tomalok** [comment]
Blocked by https://github.com/hashicorp/packer/issues/7358

**Comment by tomalok** [comment]
As noted on hashicorp/packer#7358, I've used a build of `packer` that allows the setting of `ami_architecture` to successfully build an AMI.

Granted, that AMI didn't boot properly, but I believe that's due to my not having `grub-efi` set up correctly just yet...

**Comment by thresheek** [comment]
Hey @tomalok! Thanks for your work on making aarch64 ami a thing.  Indeed, we have the same issue as in AMI doesnt boot up properly; I was wondering if that is indeed a grub issue as it seems to be set up; or maybe a kernel one?

Any idea on how could we help debugging the issue?

Thanks

**Comment by tomalok** [comment]
I haven't had spare cycles in nearly a year to dig deeper into this particular issue.

The last thing I recall is considering giving up on `grub`, and instead using `u-boot` sysloader, which apparently is capable of using the same files as `syslinux` -- making that switch is probably the next step.

**Comment by martin-g** [comment]
I would be glad to see this issue resolved!

**Comment by tsarna** [comment]
a1 was little disappointing, but with m6g coming I think this gets more interesting. 

If you enabled something like Issuehunt for this repo, I'd be willing to kick in a few bucks to help move this along.

**Comment by tsarna** [comment]
m6g family is now GA and is _very_ attractive, offering better performance than m5 at prices lower than t3 (and barely above t3a).

I renew my offer of, say, a $50 reward as a small incentive to anyone who wants to work on this.

There are already images of Red Hat, Ubuntu, SUSE, and Amazon Linux for aarch64. It's probably worth investigating what boot process they use.

**Comment by mcrute** [comment]
@tsarna I'm really interesting in getting ARM up and running as well. I've started to dig in and I think I might have a path forward.

Thanks for your offer of a bounty. I do appreciate that you're willing to put money behind an issue that's important to you but making money is not a goal for this project.

I'm not going to have time to look at this till Monday but if my research is correct I may be able to get this released for ARM next week. I'll provide an update here by the end of the week either way.

**Comment by mcrute** [comment]
I was able to work on this a little bit this week and I think I'm making progress. I don't yet have a bootable image, there are still issues making it through the boot loader. I've posted my current, heavily work in progress, branch in case anyone wants to tinker with this over the weekend. Otherwise I'll pick it back up on Tuesday and report back at the end of next week with progress.

**Comment by mcrute** [comment]
I'm happy to report that I've had some success getting a system to boot through userland. I think the only real issue is network initialization, but I have a few ideas on how to fix that. Once that's wrapped up I'll release a beta AMI on this issue and let it soak for a few days before a full release next week. I'm hoping to wrap this up today for a beta release tomorrow.

**Comment by mcrute** [comment]
Alpine isn't building the ENA driver in their kernel for aarch64. I've submitted a [patch to upstream](https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/8532) but this will be necessary for a real beta.

**Comment by kaniini** [comment]
I don't think that the AMI should be based on `linux-edge` kernel.  We are using that kernel for testing, and stability (in terms of functionality and enabled features) is not guaranteed.  Hince why it is in testing.

**Comment by mcrute** [comment]
I discussed with @kaniini in IRC but for the sake of documentation... we don't support the `linux-edge` packages and do not plan to ship them in any AMI we publicly release. The change to edge is primarily to keep lts and virt kernels in sync with edge and to not block users building derivative AMIs from using it if they have a use-case.

**Comment by mcrute** [comment]
The ENA MR for upstream noted above does allow the system to fully boot to SSH access. Now I just need to work the fixes back into the build script and profile system and this should be ready for a beta release, maybe even today. Upstream is pushing the kernel changes out into edge and should they should land in 3.12 (thanks @ncopa!).

**Comment by mcrute** [comment]
I have a working test AMI in us-west-2 for both `a1` and `m6g`. Please give it a try and report back if you run into any issues. Do **NOT** use this AMI for anything but short-lived testing as I will be deleting it within the next week. This is not a generally available ARM AMI, it's just for validation testing!

`ami-02c73acf61ee8cf92`

I'm also doing some testing and will wait for feedback before releasing an updated `edge` AMI early next week. The first stable production AMI will likely be the 3.12 series.

**Comment by tomalok** [comment]
> I discussed with @kaniini in IRC but for the sake of documentation... we don't support the `linux-edge` packages and do not plan to ship them in any AMI we publicly release. The change to edge is primarily to keep lts and virt kernels in sync with edge and to not block users building derivative AMIs from using it if they have a use-case.

there have been a couple of RCs for Alpine 3.12, hopefully the aarch64 config tweak will make it in before that happens.  i thought I'd submitted a patch to get all -virt flavors in line some time ago, but apparently not, or it's been undone in the interim.

**Comment by thresheek** [comment]
Thanks!

I've tested building 3.12 on aarch64 using the current master source and they boot just fine on a1.medium.

**Comment by mcrute** [comment]
@thresheek thanks for the feedback! I'm wrapping up a few little tooling changes and we should have the 3.12 release out for all architectures early this week.

**Comment by emiliofernandes** [comment]
Thank you for your work, @mcrute !

**Comment by martin-g** [comment]
@mcrute I guess you've been busier lately. Any idea when you will be able to release the new AMI ?

**Comment by mcrute** [comment]
@martin-g my apologies, things have been a little crazy lately. The big blocker on this is getting the release tool tested. I have it pretty much written but need to do some testing before I can get the latest images for ARM released. I'm hoping to have this done either by the end of this week or by the middle of next week. I'll update this issue on Friday either way to let everyone know what's happening. Thanks for your patience!

**Comment by martin-g** [comment]
Thank you for the update, @mcrute !

**Comment by martin-g** [comment]
hey @mcrute !
I know it is holiday season! I just want to ask you whether you have any update on this issue!

**Comment by mcrute** [comment]
@martin-g my apologies, I've gotten tied up in some other stuff and wasn't able to focus on this. The main blocker was a new release tool for opt-in regions which is now ready for review (#83). This should unblock the 3.12 release and we can move forward once that's merged.

**Comment by martin-g** [comment]
Awesome! Thank you, @mcrute !

**Comment by tsarna** [comment]
The t4g family was released yesterday and makes me even more excited for this!

**Comment by tomalok** [comment]
We recently built and [released Alpine 3.12.1](https://github.com/mcrute/alpine-ec2-ami/blob/master/releases/README.md) AMIs a couple days ago.

In addition to aarch64 AMIs, these new images also fully support IMDSv2 and hotpluggable ENIs/IPv6/secondary IPv4s.
